gensim/models/sent2vec_inner.pyx

```diff
@@ -0,0 +1,445 @@
+        Vector of word ids of context words.
+    target : const int
+        Word id of target word.
+    lr : REAL_t
+        Current learning rate.
+    grad : REAL_t *
+        Pointer to gradient vector.
+    wo : REAL_t *
+        Pointer to output sentence vector.
+    wi : REAL_t *
+        Pointer to input sentence vector.
+    hidden : REAL_t *
+        Pointer to hidden vector.
+    vector_size : const int
+        Size of the sentence embeddings.
+    neg : const int
+        Number of noise words to be drawn for negative sampling.
+    negatives : int *
+        Pointer to array containing noise words.
+    negpos : int *
+        Pointer to position of noise word to be drawn.
+    negatives_len : const int
+        Length of the array containing noise words.
+    Returns
+    -------
+    loss : REAL_t
+        Model loss.
+
+    """
+    if context.size() <= 0:
+        return 0
+
+    cdef REAL_t alpha = ONEF / <REAL_t>(context.size())
+    cdef REAL_t loss
+    cdef int i
+
+    for i from 0 <= i < context.size():
+        our_saxpy(&vector_size, &ONEF, &wi[context[i]], &ONE, hidden, &ONE)
+    sscal(&vector_size, &alpha, hidden, &ONE)
+    loss = negative_sampling(target, lr, grad, wo, hidden, vector_size, neg, negatives, negpos, negatives_len)
+    sscal(&vector_size, &alpha, grad, &ONE)
+    for i from 0 <= i < context.size():
+        our_saxpy(&vector_size, &ONEF, grad, &ONE, &wi[context[i]], &ONE)
+    return loss
+
+
+cdef REAL_t random_uniform()nogil:
+    """Generate random real number between 0 and 1.
+
+    Returns
+    -------
+    REAT_t
+        Generate random real number between 0 and 1.
+
+    """
+    return rand() / (RAND_MAX + 1.0)
+
+
+cdef int random_range(int a, int b)nogil:
+    """Generate random integer for given input range.
+
+    Parameters
+    ----------
+    a : int
+        Lower limit of input range.
+    b: int
+        Upper limit of input range.
+    Returns
+    -------
+    int
+        Random integer in given input range [a,b].
+
+    """
+    return a + <int>(rand() % ((b - a) + 1))
+
+
+cdef int get_line(vector[int] &wids, vector[int] &words, int max_line_size)nogil:
+    """Converting sentence to a list of word ids inferred from the dictionary.
+
+    Parameters
+    ----------
+    wids : vector[int]
+        Vector of word ids for a sentence.
+    words : vector[int] &
+        Reference to vector of valid word ids for a sentence.
+    max_line_size : int
+        Maximum length of input sentence.
+    Returns
+    -------
+    ntokens : int
+        Number of tokens processed in given sentence.
+
+    """
+    cdef int ntokens = <int> 0
+    cdef int i
+
+    for i from 0 <= i < wids.size():
+        if wids[i] < 0:
+            continue
+        ntokens += 1
+        words.push_back(wids[i])
+        if ntokens >= max_line_size:
+            break
+    return ntokens
+
+
+cdef void add_ngrams_train(vector[int] &line, int n, int k, int bucket, int size)nogil:
```

Comment:
shouldn't empty space stand before nogil ?

Reply:
Not necessary as I remember

Reply:
yeah, unnecessary, but r e a d a b i l i t y

Reply:
I'll fix that myself of course